Fix handling of array constructors in Crossgen2

trylek opened the pull request:
Fixes: https://github.com/dotnet/runtime/issues/48204

I have attempted to adapt Michal's fix https://github.com/dotnet/runtimelab/pull/586 to Crossgen2.

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix handling of array constructors in Crossgen2

Fixes: https://github.com/dotnet/runtime/issues/48204

sandreenko commented:
/azp run runtime-coreclr crossgen2 outerloop

trylek committed:
Ramping up on the method descriptor handle in CoreCLR

MichalStrehovsky commented:
I think we'll need to very carefully only do this for array constructors - arrays have IL-based methods on them that I think still need to canonicalize as usual. Only the constructor is special magic because it doesn't actually have code associated with it.

If you're set up for it, it might be worth checking how CoreCLR/crossgen1 does the canonical forms of these constructors (I know we try to mirror how crossgen1 is structured and crossgen1 apparently can get away without this special casing).

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -358,7 +358,11 @@ private IMethodNode CreateMethodEntrypoint(TypeAndMethod key)
             bool isPrecodeImportRequired = key.IsPrecodeImportRequired;
-            MethodDesc compilableMethod = method.Method.GetCanonMethodTarget(CanonicalFormKind.Specific);
+            MethodDesc compilableMethod = method.Method;
+            if (!compilableMethod.OwningType.IsArray)
```
I wouldn't expect special casing outside corinfoimpl to be needed - RyuJIT should realize this is an array constructor and ask for a helper - I wouldn't expect there would be an entrypoint for the actual method.

trylek replied:
Thanks Michal for your insightful feedback. I have tracked this down to the fact that we need to set the <code>codePointerOrStubLookup.constLookup</code> in the "zapinfo" version of getCallInfo, otherwise it unconditionally requested the entrypoint. As of the 2nd commit I've been able to roll this back.

trylek committed:
Improve the change based on Michal's PR feedback

Thanks to Michal's feedback I've been able to better grasp the
problem and fix deficiencies of my first commit. I have moved
setting of the constLookup to the "zapinfo version of getCallInfo",
I have modified the code to avoid canonicalization only for
array constructors and I rolled back the change to
CreateMethodEntrypoint that is no longer necessary after the
"zapinfo" fix.

Thanks

Tomas

trylek commented:
@MichalStrehovsky - this is what I see when I run the test with CG1:

<pre>
Microsoft (R) CoreCLR Native Image Generator - Version 6.0.0-dev
Copyright (c) Microsoft Corporation.  All rights reserved.

Method not found: 'Void System.Int32[]..ctor(Int32, Int32)'. while compiling method <Module>.alloc
Native image D:\git\runtime7\artifacts\tests\coreclr\windows.x64.Checked\Regressions\coreclr\Runtimelab_578\arr\\arr.dll generated successfully.
</pre>

Does that answer your question?

MichalStrehovsky commented:
The warning is for these lines:

https://github.com/dotnet/runtime/blob/49cfb3507c586b539d0652814defbbc9e3073f16/src/tests/Regressions/coreclr/Runtimelab_578/arr.il#L17-L22

That's expected since since single-dimensional arrays or non-arrays don't have a jagged constructor.

My question is more about what crossgen does on the same line that was crashing in crossgen2 - why does the canonicalization succeed. But no need to spend time on it if you're not set up for it (don't have things built). More just curiosity on my side.

mangod9 commented:
@trylek is this PR ready to merge or still needs work?

trylek commented:
Apart from the PR question I'm not aware of any outstanding blocking feedback; it may also need rebasing. I'll look into finishing that today or tomorrow.

trylek commented:
Taking a look now, will roll this back if I don't manage to nail this down and fix quickly.
